### PR TITLE
feat(runtime): Wasmer 1.0 runner

### PR DESCRIPTION
**Pull request by ailisp**
Add Wasmer 1.0 runner, enabled by wasmer1_vm features (similar to wasmtime_vm). Expose wasmer 1.0 runner and wasmtime runner available to neard and vm standalone.

Test Plan
---------
All near-vm-runner tests pass. compatible behavior as wasmer 0.17. neard compiles and work with wasmer 1.0 runner turns on

### CODE REVIEW

File: runtime/near-vm-logic/src/config.rs
Comment (olonho) [inline runtime/near-vm-logic/src/config.rs:11]: Maybe rename to Wasmer0.

File: runtime/near-vm-runner/tests/test_error_cases.rs
Comment (frol) [inline runtime/near-vm-runner/tests/test_error_cases.rs:290]: Just a friendly reminder :smile:

**Review by frol — APPROVED** [review]

**Review by bowenwang1996 — APPROVED** [review]

File: runtime/near-vm-errors/src/lib.rs
Comment (ailisp) [reply runtime/near-vm-errors/src/lib.rs:43]: Talked with Wasmer team, it is not that bad now. They have three kind of RuntimeError: Trap, User and General. User kind can be detected and downcast to VMLogicError (implemented in this pr). The other two (indicate what kind of trap it is) are still not, they suggest me to create a pr to expose them

File: runtime/near-vm-runner/src/imports.rs
Comment (ailisp) [reply runtime/near-vm-runner/src/imports.rs:12]: 👍 I'll reuse ImportReference and delete this struct. they were different but is same now

File: runtime/near-vm-errors/src/lib.rs
Comment (ailisp) [reply runtime/near-vm-errors/src/lib.rs:43]: I made a pr to expose it and it's merged, will use git version of wasmer for now

Comment (ailisp) [reply runtime/near-vm-errors/src/lib.rs]: It's fixed with git wasmer

**Comment by ailisp** [comment]
@olonho Finished, please take another look
@khorolets need your code owner review because a version confliction, need smallvec 1.6 <--- actix 0.10

File: runtime/near-vm-runner/Cargo.toml
Comment (willemneal) [inline runtime/near-vm-runner/Cargo.toml:19]: Is 1.0 not release to crates.io yet?

Comment (ailisp) [reply runtime/near-vm-runner/Cargo.toml:19]: It's released but i need a patch https://github.com/wasmerio/wasmer/pull/2026, it might be released in 1.0.2

**Review by willemneal — APPROVED** [review]
Just my one question about the dep version. Otherwise LGTM!

File: tools/indexer/example/Cargo.toml
Comment (ailisp) [reply tools/indexer/example/Cargo.toml:8]: Yes, wasmer 1.0.1 ask for smallvec 1.6, and so does actix 0.10. With actix 0.9 it's smallvec 1.5 and cause cargo confliction. I looked up solutions and it is said if a crate is rust, it can have multiple version, but if it has c deps, it cannot, so easiest fix is bump actix

**Review by khorolets — APPROVED** [review]

File: runtime/near-vm-runner/Cargo.toml
Comment (olonho) [reply runtime/near-vm-runner/Cargo.toml:19]: Shalln't we ask them to publish it and depend on published version?

File: runtime/near-vm-runner/src/imports.rs
Comment (olonho) [inline runtime/near-vm-runner/src/imports.rs:12]: MyEnv is not the best name, maybe NearWasmerEnv?

File: runtime/near-vm-runner/src/wasmer1_runner.rs
Comment (olonho) [inline runtime/near-vm-runner/src/wasmer1_runner.rs:144]: Significant amount of copy-paste, but not sure what could be done here.

**Review by olonho — COMMENTED** [review]
Overall looks good, couple naming issues. and maybe depend on published Wasmer crate.

File: runtime/near-vm-runner/src/wasmer1_runner.rs
Comment (ailisp) [reply runtime/near-vm-runner/src/wasmer1_runner.rs:144]: Yeah run_wasmer1 is structurally very alike to run_wasmer, but every step a different type of wasmer instance, module, etc. is used between 0.x and 1.x, so it's difficult to combine them

**Comment by ailisp** [comment]
@olonho @evgenykuzyakov please approve to unblock :D

**Review by olonho — APPROVED** [review]
LGTM

**Review by olonho — APPROVED** [review]

File: tools/indexer/example/Cargo.toml
Comment (bowenwang1996) [reply tools/indexer/example/Cargo.toml:8]: why is this not added to `deny.toml`?

File: runtime/near-vm-runner/src/imports.rs
Comment (frol) [inline runtime/near-vm-runner/src/imports.rs:7]: 1. Is it safe to clone the pointer? Who is cleaning the memory?
2. Can we avoid `pub`? Would `pub(crate)` be enough?

File: runtime/near-vm-runner/src/wasmer1_runner.rs
Comment (frol) [inline runtime/near-vm-runner/src/wasmer1_runner.rs:159]: :+1:

File: runtime/near-vm-runner/tests/test_error_cases.rs
Comment (frol) [inline runtime/near-vm-runner/tests/test_error_cases.rs:282]: Is it safe to do parallel testing of Wasmer 0.17 and Wasmer 1.0 given the context this comment provides? I wonder if we have any protection in our regular (non-testing) context against parallel execution of VMs.

File: tools/indexer/example/Cargo.toml
Comment (frol) [reply tools/indexer/example/Cargo.toml:8]: @bowenwang1996 it seems that the nature of the issue @ailisp talks about is not related to organizational dependency duplication, rather a technical limitation of C libraries that collide in their exported symbols. I am not sure how it is related to pure-rust `smallvec` crate, though. @ailisp Did you bump into some funny compilation/runtime error or was it a failure from `cargo-deny` on CI?

Comment (ailisp) [reply tools/indexer/example/Cargo.toml:8]: @frol the compilation error message was "unable to find a version of smallvec that satisfy blah blah", however it's not reproducible now. I had clear cargo cache since last time, maybe cargo cache scammed somehow?

Looking a little into Cargo.lock:
```
[[package]]
name = "actix"
version = "0.9.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "a4af87564ff659dee8f9981540cac9418c45e910c8072fdedd643a262a38fcaf"
dependencies = [
 ...
 "smallvec 1.6.1",
]
```

^ used to resolve to smallvec 1.5.x, but now it also becomes 1.6.1, so there is only smallvec 1.6.1 in cargo.lock, thus no compilation error 🤔

Comment (ailisp) [reply tools/indexer/example/Cargo.toml]: I undo actix to 0.9 since this compilation error doesn't exist now

File: runtime/near-vm-runner/Cargo.toml
Comment (ailisp) [reply runtime/near-vm-runner/Cargo.toml:19]: Asked

File: runtime/near-vm-runner/src/imports.rs
Comment (ailisp) [reply runtime/near-vm-runner/src/imports.rs:7]: 1. Yes. The pointer is convert from VMLogic object. clone is to clone VMLogic object into wasmer memory. Wasmer does the cleaning on wasm runner exit.
2. No. Because this macro:
```
error[E0446]: crate-visible type `NearWasmerEnv` in public interface
   --> runtime/near-vm-runner/src/imports.rs:59:17
    |
13  |   pub(crate) struct NearWasmerEnv {
    |   ---------- `NearWasmerEnv` declared as crate-visible
...
59  |                   pub fn $func(env: &NearWasmerEnv, $( $arg_name: $arg_type ),* ) -> VMResult<($( $returns ),*)> {
    |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can't leak crate-visible type
...
175 | / wrapped_imports! {
176 | |     // #############
177 | |     // # Registers #
178 | |     // #############
...   |
319 | |     // #["protocol_feature_evm", EVM] test_api<[a: u64] -> []>,
320 | | }
    | |_- in this macro invocation
```
require NearWasmer and its fields to be pub. And therefore `    pub logic: ImportReference,` cannot be pub(crate)` because of "   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^ can't leak crate-visible type"

Comment (ailisp) [reply runtime/near-vm-runner/src/imports.rs:7]: This actually is implicitly "pub(crate)" because it's in `mod imports;` which is not pub, therefore other crate cannot import it.

File: runtime/near-vm-runner/tests/test_error_cases.rs
Comment (ailisp) [reply runtime/near-vm-runner/tests/test_error_cases.rs:282]: It is safe, as the comment said wasmtime signals would be falsely caught by wasmer, this is due to specialty of wasmtime throw signals. Also it's not this test failed, it's other test execute in parallel incorrectly catch wasmtime signals. So just avoid run wasmtime signal test in this test case will not break our test

File: runtime/near-vm-runner/Cargo.toml
Comment (evgenykuzyakov) [inline runtime/near-vm-runner/Cargo.toml:18]: Should this also be optional now and be enabled only by `wasmer_default`?

File: runtime/near-vm-runner/src/cache.rs
Comment (evgenykuzyakov) [inline runtime/near-vm-runner/src/cache.rs:226]: This can be much easier if we moved VM specific caching into separate mods. E.g. you have mod (as a folder) `wasmer_0_17` and mod `wasmer_1`. They have the same interface and you only need to import them once. Instead of having to mark each method with `#[cfg(feature = "wasmer1_vm")]`

File: runtime/near-vm-runner/src/errors.rs
Comment (evgenykuzyakov) [inline runtime/near-vm-runner/src/errors.rs:6]: Why not use `std::Into` Trait?

File: runtime/near-vm-runner/src/imports.rs
Comment (evgenykuzyakov) [inline runtime/near-vm-runner/src/imports.rs:141]: Are you missing protocol specific features here?

E.g. from wasmer0.17
```
$(#[cfg(feature = $feature_name)])*
if true $(&& near_primitives::checked_feature!($feature_name, $feature, protocol_version))* {
    ns.insert(stringify!($func), wasmer_runtime::func!(wasmer_ext::$func));
}
```

File: runtime/near-vm-runner/src/runner.rs
Comment (evgenykuzyakov) [inline runtime/near-vm-runner/src/runner.rs:234]: Why?

File: runtime/near-vm-runner/src/runner.rs
Comment (evgenykuzyakov) [inline runtime/near-vm-runner/src/runner.rs:275]: wrong feature?

File: runtime/near-vm-runner/src/wasmer1_runner.rs
Comment (evgenykuzyakov) [inline runtime/near-vm-runner/src/wasmer1_runner.rs:128]: Is unwrap safe here?

File: runtime/near-vm-runner/src/wasmer1_runner.rs
Comment (evgenykuzyakov) [inline runtime/near-vm-runner/src/wasmer1_runner.rs:249]:
```suggestion
            Err(e) => (Some(logic.outcome()), Some(e.into_vm_error())),
        },
        Err(e) => (Some(logic.outcome()), Some(e.into_vm_error())),
```

**Review by evgenykuzyakov — CHANGES_REQUESTED** [review]

File: runtime/near-vm-runner/src/cache.rs
Comment (ailisp) [reply runtime/near-vm-runner/src/cache.rs:226]: Good idea

File: runtime/near-vm-runner/src/runner.rs
Comment (ailisp) [reply runtime/near-vm-runner/src/runner.rs:234]: added

**Review by evgenykuzyakov — APPROVED** [review]
Please address nit comments first

File: runtime/near-vm-errors/src/lib.rs
Comment (ailisp) [reply runtime/near-vm-errors/src/lib.rs:317]: There're two calls this with `e.into()`, one `e` is a &VMLogicError, another is VMLogicError, so make a `From<&VMLogicError>` works for both

File: runtime/near-vm-runner/src/wasmer1_runner.rs
Comment (ailisp) [reply runtime/near-vm-runner/src/wasmer1_runner.rs:128]: It is safe, this function is almost identical to wasmer_runner::check_method, just `wasmer_runtime_core::module::ExportIndex::Func` => `wasmer_types::ExportIndex::Function` refactor done by wasmer. as i checked wasmer code, when `Function(index)` is in `module.info().exports`, corresponding `index` always exist in `info.functions`, say it is `func`, and `func` is always in `info.signatures`

**Pull request merged**
